### PR TITLE
Feature/ui stylesheets-enhancements

### LINKED ISSUE #159: ui sylesheets

### PR DESCRIPTION
**Pull request by goldsborough**
closes #159 
This PR brings two main changes to the UI theming code (or changes *for* it):

1. Refactoring of the `ProjectCreationScreen` into one file per section.
2. Creation of a `Status` and `StatusWithValue` class to propagate "exceptions" in Qt without needing to compile the Qt binary with exceptions. This is needed to catch very early errors that would prevent starting the UI because of some basic problems, such as the theme (or, later, configuration) folder not being found. By "throwing" an exception (returning error codes) we can catch such problems before loading the main QML file and instead loading only an error dialog showing the problem.

You can try it out by making a spelling mistake in `ui.cpp`, line 57:

All OK:
```cpp
auto status = Theme::Make("default");
```

Typo causes error box to show:
```cpp
auto status = Theme::Make("banana"); // minor typo
```

### CODE REVIEW

File: include/common/status.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  /**
+   * \returns The message of the status.
+   */
+  const std::string& message() const noexcept;
+
+  /**
+   * \returns True if the status code is OK, else false.
+   */
+  explicit operator bool() const noexcept {
+    return _code == Code::OK;
```
Comment (janschopohl) [inline include/common/status.hpp]: -> cpp

File: include/common/status-with-value.hpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  const T value() const noexcept {
+    assert::that(static_cast<bool>(_value));
+    return *_value;
+  }
+
+  /**
+   * \returns The status.
+   */
+  const Status& status() const noexcept {
+    return _status;
```
Comment (janschopohl) [inline include/common/status-with-value.hpp:90]: I guess this should be in the cpp file, for better style.

File: include/ui/theme.hpp [hunk 3]
```diff
@@ -52,24 +54,41 @@ class Theme : public QQmlPropertyMap {
...
+   *
+   * This factory function should only be called once.
+   * The class does not take ownership of the allocated theme, it
+   * only provides access to it.
+   *
+   * \param themeName The theme to initialize the new theme singleton with.
    * \returns A pointer to the newly created theme.
    */
-  static Theme* reset(const QString& themeName, QObject* parent = nullptr);
+  static Status Make(const QString& themeName);
```
Comment (janschopohl) [inline include/ui/theme.hpp:66]: Is there a reason for uppercase?

**Review by janschopohl — APPROVED** [review]
I tested this quite a bit and everything worked for me.
👍

File: include/ui/theme.hpp (hunk 3 above)
Comment (goldsborough) [reply include/ui/theme.hpp:66]: It's a factory function (see also `NodeFactoryCollection`)

File: include/common/status-with-value.hpp (hunk 2 above)
Comment (goldsborough) [reply include/common/status-with-value.hpp:90]: Unfortunately it's a template :(

**Pull request merged**